sources/packages/backend/apps/queue-consumers/src/services/cas-invoice-batch/cas-invoice-batch.service.ts

```diff
@@ -0,0 +1,254 @@
+ */
+const INVOICES_CHUNK_SIZE_INSERTS = 150;
+
+@Injectable()
+export class CASInvoiceBatchService {
+  constructor(
+    private readonly dataSource: DataSource,
+    private readonly sequenceControlService: SequenceControlService,
+    private readonly systemUsersService: SystemUsersService,
+  ) {}
+
+  /**
+   * Find e-Cert receipts that have no valid invoice associated,
+   * creating a new invoice batch to be sent to CAS if some
+   * pending receipts are found.
+   * @param processSummary process summary for logging.
+   * @returns new invoice batch and related data or null.
+   */
+  async createInvoiceBatch(
+    processSummary: ProcessSummary,
+  ): Promise<CASInvoiceBatch | null> {
+    return this.dataSource.transaction(async (entityManager) => {
+      let newBatchUniqueSequence: number;
+      // Consumes the next sequence number and prevent concurrent access.
+      await this.sequenceControlService.consumeNextSequenceWithExistingEntityManager(
+        "CAS_INVOICE_BATCH",
+        entityManager,
+        async (nextSequenceNumber: number) => {
+          newBatchUniqueSequence = nextSequenceNumber;
+        },
+      );
+      processSummary.info("Checking for pending receipts.");
+      const pendingReceipts = await this.getPendingReceiptsForInvoiceBatch(
+        entityManager,
+      );
+      if (!pendingReceipts.length) {
+        processSummary.info("No pending receipts found.");
+        // Cancels the transaction to rollback the batch number.
+        await entityManager.queryRunner.rollbackTransaction();
+        return null;
+      }
+      processSummary.info(`Found ${pendingReceipts.length} pending receipts.`);
+      const now = new Date();
+      // Create the new invoice batch.
+      const invoiceBatch = new CASInvoiceBatch();
+      invoiceBatch.batchName = `SIMS-BATCH-${newBatchUniqueSequence}`;
+      invoiceBatch.batchDate = now;
+      invoiceBatch.approvalStatus = CASInvoiceBatchApprovalStatus.Pending;
+      invoiceBatch.approvalStatusUpdatedOn = now;
+      invoiceBatch.approvalStatusUpdatedBy = this.systemUsersService.systemUser;
+      invoiceBatch.creator = this.systemUsersService.systemUser;
+      // Save the invoice batch to allow the ID to be known and associate with every new
+      // invoice created. It would allow the invoices to be saved in chunks.
+      await entityManager.getRepository(CASInvoiceBatch).save(invoiceBatch);
+      // Accumulates the invoices to be returned later.
+      const casInvoices: CASInvoice[] = [];
+      // Get active distribution accounts to be associated with each invoice detail.
+      const distributionAccounts = await this.getActiveDistributionAccounts(
+        entityManager,
+      );
+      await this.sequenceControlService.consumeNextSequenceWithExistingEntityManager(
+        "CAS_INVOICE",
+        entityManager,
+        async (nextSequenceNumber: number) => {
+          let newInvoiceSequence = nextSequenceNumber;
```

Comment:
I would strongly suggest moving line 86-127 into a private method. IMO it would help the readability.

Reply:
I refactored to extract the method, please take a look.

Reply:
Thanks.